### PR TITLE
#341: Added support for step debugger and memory dumps after each op invocation

### PR DESCRIPTION
**Pull request by tapspatel**
2 new modes in ttrt (debugger + memory)
```
ttrt run <flatbuffer> --debugger
```
This will start a pdb debugger after every op invocation in runtime. Further support will be added.

```
ttrt run <flatbuffer> --memory
ttrt perf <flatbuffer> --memory
ttrt run/perf <flatbuffer> --memory --check-memory-leak
```

This will dump memory reports after each op invocation in runtime. A memory_report.json file will be dumped, containing all the ops (an entry for each) and the dram/l1 memory usage. This is global view of the board.

```
{
        "loc": "loc(\"/code/tt-mlir/test/python/golden/test_ttir_ops.py:65:id(0)\")",
        "debug_str": "%6 = \"ttnn.add\"(%2, %4, %5) <{operandSegmentSizes = array<i32: 2, 1>}> : (tensor<64x128xf32, #ttnn.ttnn_layout<(d0, d1) -> (d0, d1), <1x1>, memref<2x4x!tt.tile<32x32, f32>, #ttnn.buffer_type<dram>>, <interleaved>>>, tensor<64x128xf32, #ttnn.ttnn_layout<(d0, d1) -> (d0, d1), <1x1>, memref<2x4x!tt.tile<32x32, f32>, #ttnn.buffer_type<dram>>, <interleaved>>>, tensor<64x128xf32, #ttnn.ttnn_layout<(d0, d1) -> (d0, d1), <1x1>, memref<2x4x!tt.tile<32x32, f32>, #ttnn.buffer_type<dram>>, <interleaved>>>) -> tensor<64x128xf32, #ttnn.ttnn_layout<(d0, d1) -> (d0, d1), <1x1>, memref<2x4x!tt.tile<32x32, f32>, #ttnn.buffer_type<dram>>, <interleaved>>> loc(\"/code/tt-mlir/test/python/golden/test_ttir_ops.py:65:id(0)\")",
        "dram": {
            "total_allocatable (bytes) : total_allocatable/bank * num_banks": "12884901504",
            "total_allocated (bytes) : total_allocated/bank * num_banks": "3268608",
            "total_free (bytes) : total_allocatable - total_allocated": "12881632896",
            "blocks": [
                {
                    "address (bytes)": "32",
                    "size (bytes)": "90112",
                    "allocated (y/n)": "Y"
                },
                {
                    "address (bytes)": "90144",
                    "size (bytes)": "90112",
                    "allocated (y/n)": "Y"
                },
                {
                    "address (bytes)": "180256",
                    "size (bytes)": "4096",
                    "allocated (y/n)": "N"
                },
                {
                    "address (bytes)": "184352",
                    "size (bytes)": "90112",
                    "allocated (y/n)": "Y"
                },
                {
                    "address (bytes)": "274464",
                    "size (bytes)": "1073465312",
                    "allocated (y/n)": "N"
                },
                {
                    "address (bytes)": "1073739776",
                    "size (bytes)": "2048",
                    "allocated (y/n)": "Y"
                }
            ],
            "total_allocatable (bytes) : per bank": "1073741792",
            "total_allocated (bytes): per bank": "272384",
            "total_free (bytes) : per bank": "1073469408",
            "largest_free_block (bytes) : per bank": "1073465312"
        },
        "l1": {
            "total_allocatable (bytes) : total_allocatable/bank * num_banks": "87504896",
            "total_allocated (bytes) : total_allocated/bank * num_banks": "0",
            "total_free (bytes) : total_allocatable - total_allocated": "87504896",
            "blocks": [
                {
                    "address (bytes)": "99104",
                    "size (bytes)": "1367264",
                    "allocated (y/n)": "N"
                }
            ],
            "total_allocatable (bytes) : per bank": "1367264",
            "total_allocated (bytes): per bank": "0",
            "total_free (bytes) : per bank": "1367264",
            "largest_free_block (bytes) : per bank": "1367264",
            "largest_contiguous_free_block (bytes) : per bank": "1367264"
        }
    },
```

### CODE REVIEW

File: runtime/tools/python/ttrt/common/debugger.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    device = device_runtime_config.device
+
+    print("-----------executing debugger-----------")
+
+    try:
+        import pdb
+
+        pdb.set_trace()
+    finally:
+        print("-----------finished executing debugger-----------")
```
Comment (nsmithtt) [inline runtime/tools/python/ttrt/common/debugger.py:40]: I think we should remove these print statements

File: runtime/tools/python/ttrt/common/memory.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        device.dump_memory_report()
+
+        op_debug_str = ttrt.runtime.get_op_debug_str(op_context)
+
+        # find matching golden tensor based on loc in op debug string
+        match = re.search(r"loc\(([^)]+)\)", op_debug_str)
+
+        if not match:
+            print(f"debug_str={op_debug_str}")
+            print("No location found in debug string - skipping memory dump")
```
Comment (nsmithtt) [inline runtime/tools/python/ttrt/common/memory.py:49]: Maybe the prints should be logger statements? It's unclear that we want to always spew this information, I could imagine it being very noisy for big models with lots of ops

File: runtime/tools/python/ttrt/common/run.py [hunk 3]
```diff
@@ -150,7 +152,7 @@ def initialize_api():
         Run.register_arg(
             name="--golden",
             type=bool,
-            default=True,
+            default=False,
```
Comment (nsmithtt) [inline runtime/tools/python/ttrt/common/run.py:155]: Intentional? Seems unrelated to this change

File: runtime/tools/python/ttrt/common/memory.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    print("-----------executing memory dump-----------")
+
+    try:
+        import ttrt.runtime
+        import ttrt.binary
+
+        device.dump_memory_report()
+
+        op_debug_str = ttrt.runtime.get_op_debug_str(op_context)
```
Comment (nsmithtt) [inline runtime/tools/python/ttrt/common/memory.py:42]: Should we just use `get_op_loc_info` instead of trying to parse the loc out of the debug string?

File: runtime/tools/python/ttrt/common/memory.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        memory_dump_dir_path = (
+            f"{memory_runtime_config.tt_metal_home_path}/generated/reports"
+        )
+        memory_dump_artifact_dir_path = (
+            f"{memory_runtime_config.artifact_dir}/{row_col}_memory_reports"
+        )
+        # os.mkdir(memory_dump_artifact_dir_path)
+        if os.path.exists(memory_dump_artifact_dir_path):
+            shutil.rmtree(memory_dump_artifact_dir_path)
+        shutil.copytree(memory_dump_dir_path, memory_dump_artifact_dir_path)
```
Comment (nsmithtt) [inline runtime/tools/python/ttrt/common/memory.py:73]: Instead of moving dir trees around, maybe it's better to read the memory dump file and bring it into a common top level json file that's easier for us to use?

That way we can also just encode the loc / debug strings intact in the json instead of relying on row/col naming conventions.

**Comment by odjuricicTT** [comment]
@tapspatel Will `--memory` be available with `ttrt perf` as well?

File: runtime/tools/python/ttrt/common/debugger.py (hunk 1 above)
Comment (tapspatel) [reply runtime/tools/python/ttrt/common/debugger.py:40]: fixed

File: runtime/tools/python/ttrt/common/memory.py (hunk 2 above)
Comment (tapspatel) [reply runtime/tools/python/ttrt/common/memory.py:49]: agreed. I'll keep relevant print statements in logging.debug

File: runtime/tools/python/ttrt/common/run.py (hunk 3 above)
Comment (tapspatel) [reply runtime/tools/python/ttrt/common/run.py:155]: fixed

File: runtime/tools/python/ttrt/common/memory.py (hunk 5 above)
Comment (tapspatel) [reply runtime/tools/python/ttrt/common/memory.py:73]: agreed. will fix

File: runtime/tools/python/ttrt/common/memory.py (hunk 4 above)
Comment (tapspatel) [reply runtime/tools/python/ttrt/common/memory.py]: agree. fixed

**Comment by tapspatel** [comment]
@odjuricicTT yeah I added it to ttrt perf as well

**Review by kmabeeTT — APPROVED** [review]
Thanks, Taps!

**Comment by tapspatel** [comment]
Added `--check-memory-leak`

File: runtime/tools/python/ttrt/common/run.py [hunk 6]
```diff
@@ -563,6 +593,65 @@ def _execute(binaries):
...
+                                        if dram_final_size > dram_initial_size:
+                                            raise Exception(
+                                                "Memory leak detected in DRAM"
+                                            )
+
+                                        if l1_final_size > l1_initlal_size:
+                                            raise Exception(
+                                                "Memory leak detected in L1 cache"
+                                            )
+
```
Comment (nsmithtt) [inline runtime/tools/python/ttrt/common/run.py:654]: ** This function is getting a bit unwieldy esp with nested control flow.  This new code might be better encapsulated inside of a utility function.

File: runtime/tools/python/ttrt/common/callback.py [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+
+    if callback_runtime_config.enable_memory:
+        memory(callback_runtime_config, binary, program_context, op_context)
+
+    if callback_runtime_config.enable_debugger:
+        debugger(callback_runtime_config, binary, program_context, op_context)
+
+
+def get_callback_fn(callback_runtime_config):
+    return partial(callback, callback_runtime_config)
```
Comment (nsmithtt) [inline runtime/tools/python/ttrt/common/callback.py:440]: I'll leave this refactor up to you, but just my 2 cents:
Each of these callbacks might be better organized inside of their own subdirectories & files, `callbacks/golden.py`, `callbacks/memory.py`, `callbacks/debugger.py`.  They could also each have their own `Config` object.

You could compose them together with something like:
```python
def get_callback_fn(golden_config=None, memory_config=None, debugger_config):
    callbacks = []
    if golden_config is not None:
        callbacks.append(partial(golden, golden_config)
    if memory_config is not None:
        callbacks.append(partial(memory, memory_config)
    if debugger_config is not None:
        callbacks.append(partial(debugger, debugger_config)

    def callback_fn(*args, **kwargs):
        for fn in callbacks:
            fn(*args, **kwargs)

    return callback_fn
```

**Review by nsmithtt — APPROVED** [review]
Looks good! I added a ** to the comment I think we should address.  The other comment is optional.
